File: fsharp/pythagorean-triplet/PythagoreanTriplet.fs
module PythagoreanTriplet

let tripletsWithSum (sum: int): (int * int * int) list =
    [1..(sum/3)] 
    |> List.choose (fun a -> 
        let bs = 
            [(a+1)..(sum-a)/2] 
            |> List.filter (fun b -> 
                 a*a + b*b = (sum - a - b)*(sum - a - b))
        if bs |> List.length > 0 then Some (a, List.head bs, sum - a - List.head bs)
        else None  )

File: csharp/pythagorean-triplet/PythagoreanTriplet.cs
using System;
using System.Collections.Generic;

public static class PythagoreanTriplet
{
    public static IEnumerable<(int a, int b, int c)> TripletsWithSum(int sum)
    {
        var triplets = new List<(int, int, int)>();
        int upperLimitA = sum / 3;
        for (int a_cand = 1; a_cand <= upperLimitA; a_cand++)
        {
            int upperLimitB = (sum - a_cand) / 2;
            for (int b_cand = a_cand + 1; b_cand <= upperLimitB; b_cand++)
            {
                int c_cand = sum - a_cand - b_cand;
                if (a_cand * a_cand + b_cand * b_cand == c_cand * c_cand) triplets.Add((a_cand, b_cand, c_cand));
            }
        }
        return triplets;
    }
}